src/dependenpy/algorithms.py

```diff
@@ -0,0 +1,328 @@
+
+        1. For each element, if it does not depend on any other (empty column), move to beginning and ignore hereafter.
+        2. For each element, if nobody depends on it (empty row), move to end and ignore hereafter.
+        3. For the rest of the elements, find loops by decomposition or partition
+        4. Collapse each loop into onecompounded element, and go to step 1.
+
+        Args:
+            matrix (DsmMatrix): _description_
+        """
+        result1 = self
+        for item in range(result1.front, result1.back):
+            print("---", item, result1.front, result1.back)
+            if not self.has_inputs(item):
+                # print(">")
+                result1 = result1.element_to_back(item)
+            elif not self.has_outputs(item):
+                # print("<")
+                result1 = result1.element_to_front(item)
+            else:
+                # print("=")
+                pass
+        return result1
+
+    def decompose(self):
+        """_summary_.
+
+        Args:
+            matrix (DsmMatrix): _description_
+        """
+
+    def partition(self) -> List[Tuple[str, int]]:
+        """Builds a hierarchy of elements.
+
+        Inside a level the elements are either not connected to each other, or are part of the same circuit at that level.
+
+        The top level is composed of all elements that require no input or are independent from all other elements.
+        Removing these elements from the matrix, leaves a new matrix whose top level is the secont level of the original one.
+        In this manner, all levels can be identified.
+
+        For each element,list the set of their inputs (including the element itself), the set of the outputs
+        (including the element itself) and the intersection of these sets.
+        If the input set equals the intersection set, it is a top element.
+        After identifying all top elements, remove them and proceed recursively.
+
+        The result is an list of tuples (element, level), where the top level is 1.
+
+        Args:
+            matrix (DsmMatrix): _description_
+
+        Returns:
+            List[Tuple[str, int]]: _description_
+        """
+        return []
+
+    def _switch_rows(self, source: int, target: int) -> DsmMatrix:
+        """Returns a matrix with the given rows switched.
+
+        Args:
+            source: (int): _description_
+            target (int): _description_
+
+        Returns:
+            DsmMatrix: _description_
+        """
+        result = self.copy()
+        ii, jj = result.data[target], result.data[source]
+        result.data[source] = ii
+        result.data[target] = jj
+        return result
+
+    def _switch_columns(self, source: int, target: int) -> DsmMatrix:
+        """Returns a matrix with the given columns switched.
+
+        Args:
+            source: (int): _description_
+            target (int): _description_
+
+        Returns:
+            DsmMatrix: _description_
+        """
```

Comment:
Just a note to say that you don't have to write docstrings for private methods. It's always better to document methods of course, but if the code is easy to read and type annotated, no docstring is fine :slightly_smiling_face: